lib/tower-http-digest/src/lib.rs

```diff
@@ -0,0 +1,186 @@
+        match self {
+            Self::Sha256 => Sha256::digest(data).to_vec(),
+            Self::Sha512 => Sha512::digest(data).to_vec(),
+        }
+    }
+}
+
+pin_project! {
+    #[project = DigestFutureProj]
+    pub enum DigestFuture<S, F> {
+        BuildingDigest {
+            service: S,
+
+            algorithm: Algorithm,
+            parts: Option<http::request::Parts>,
+
+            #[pin]
+            body: Body,
+            body_accumulator: Option<BytesMut>,
+        },
+        PollServiceFuture {
+            #[pin]
+            future: F,
+        },
+    }
+}
+
+impl<S> Future for DigestFuture<S, S::Future>
+where
+    S: Service<Request<Body>>,
+    S::Error: Into<BoxError>,
+{
+    type Output = Result<S::Response, BoxError>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut task::Context<'_>) -> Poll<Self::Output> {
+        loop {
+            match self.as_mut().project() {
+                DigestFutureProj::BuildingDigest {
+                    service,
+                    algorithm,
+                    parts,
+                    mut body,
+                    body_accumulator,
+                } => {
+                    while let Some(frame) = ready!(body.as_mut().poll_frame(cx))
+                        .transpose()
+                        .map_err(BoxError::from)?
+                    {
+                        if let Ok(data) = frame.into_data() {
+                            let accumulator = body_accumulator
+                                .as_mut()
+                                .expect("[Bug] Missing accumulator");
+
+                            accumulator.put(data);
+                        }
+                    }
+
+                    let digest_header = parts.as_ref().unwrap().headers.get(&DIGEST_HEADER_NAME);
+                    let algorithm = if let Some(digest_header) = digest_header {
+                        let Some((algorithm_name, ..)) = digest_header.to_str()?.split_once('=')
+                        else {
+                            return Poll::Ready(Err("Invalid header value".into()));
+                        };
+                        Algorithm::from_str(algorithm_name)?
+                    } else {
+                        *algorithm
+                    };
+
+                    let accumulator = body_accumulator
+                        .take()
+                        .expect("[Bug] Missing accumulator")
+                        .freeze();
+
+                    let hash = algorithm.digest(&accumulator);
+                    let encoded_digest = base64_simd::STANDARD.encode_to_string(hash);
+
+                    let header_value = format!("{}={}", algorithm.as_ref(), encoded_digest);
+                    let header_value = HeaderValue::from_str(&header_value).unwrap();
+
+                    let mut parts = parts.take().expect("[Bug] Missing parts");
+                    parts.headers.insert(&DIGEST_HEADER_NAME, header_value);
```

Comment:
There actually isn't. It's just a different approach. The comparison of the digests might be the better approach even since it doesn't require a comparatively expensive call to the RSA implementation (expensive compared to just calculating a SHA digest using the rather readily available hardware acceleration; ring is pretty fast since it's based on BoringSSL).

This simply followed the approach that Akkoma and Pleroma chose (I used to be deeper into the Pleroma code way back when).

Reply:
And this would still be fine when combined with #195 since if the digests of the body and the one in the header match, and the HTTP signature check with the digest header included succeeds, then the body that was present during signing must necessarily be the same we received.